kernel/aster-nix/src/util/iovec.rs

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::mem;
+
+use super::read_val_from_user;
+use crate::{
+    prelude::*,
+    util::{read_bytes_from_user, write_bytes_to_user},
+};
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub struct IoVec {
+    base: Vaddr,
+    len: usize,
+}
+
+impl IoVec {
+    pub const fn base(&self) -> Vaddr {
+        self.base
+    }
+
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn read_from_user(&self, buffer: &mut [u8]) -> Result<()> {
+        debug_assert!(buffer.len() == self.len);
+        if buffer.len() != self.len {
+            return_errno_with_message!(
+                Errno::EINVAL,
```

Comment:
There is a `debug_assert`, but I still want to ask, do we really want to expose kernel implementation bugs as (random) user-facing error codes?

Apart from that, how do you feel about revising the APIs as following:
```rust
pub fn read_from_user<R>(&self, consumer: FnOnce(&[u8]) -> Result<R>) -> Result <R> {
    /* ... */
}

pub fn write_from_user<R>(&self, producer: FnOnce(&mut [u8]) -> Result<R>) -> Result<R> {
    /* ... */
}
```
where the `consumer` consumes the data (e.g., by calling `try_sendto`) and the `producer` produces the data (e.g., by calling `try_recvfrom`).

Reply:
> do we really want to expose kernel implementation bugs as (random) user-facing error codes?

I think we should not expose these errors to users. I have refactored the code, use `assert_eq` instead of `debug_assert`, so we don't need to return error code.

As for your suggesting API changes, it's mostly reasonable to me. But I have one concern:

These APIs suggest the kernel buffer should be allocated by `IoVec` itself, and then used by `producer` or `consumer`. but indeed, `IoVec` may not have enough information to decide which type of memory to allocate, for example, allocate heap memory, or `VmFrames`, or DMA-capable memory. Only the user knows what the exact memory type they do need.

Reply:
Thank you for your opinion. Your point makes sense to me. After more thought, I completely agree that `IoVec` should not allocate the kernel buffer.

But after more thought, I also think that the APIs of `IoVec` probably need to  be better revised in another direction to solve https://github.com/asterinas/asterinas/issues/711. Solving #711 can be outside the scope of this PR, but it has some implications on the API design.

Here, `IoVec` is essentially a slice to user memory  (containing the address to the user buffer and the size of the user buffer), so it can be directly passed to smoltcp (e.g., through the closure in [`Socket::recv`](https://docs.rs/smoltcp/latest/smoltcp/socket/tcp/struct.Socket.html#method.recv)) to detect `EFAULT`s occurred in the middle of the user buffer. If we want to do this, we need flexible APIs for the `IoVec`. For example:
 - we need different APIs for truncatable data (e.g., TCP streams)  and non-truncatable data (e.g., UDP datagrams), perhaps with methods like `read`/`write` and `read_exact`/`write_exact`.
 - we should not require the size of the kernel buffer (e.g., the size of a UDP message) to be the same as the size of the user buffer;
 - we should take `&mut self` and advance the pointer to the user message whenever necessary.

I don't think you want to do that in this PR, do you? If the answer is negative, I think we can also accept the current APIs of `IoVec` as a temporary solution to support `sendmsg` and `recvmsg`.

Reply:
Thanks for your advice!

> I don't think you want to do that in this PR, do you?

Yes, the problem seems to be out of the scope of this PR. It may require a lot of API changes and I don't want to include them in this PR.

> Here, IoVec is essentially a slice to user memory (containing the address to the user buffer and the size of the user buffer), so it can be directly passed to smoltcp (e.g., through the closure in [Socket::recv](https://docs.rs/smoltcp/latest/smoltcp/socket/tcp/struct.Socket.html#method.recv))

Quite agree. If so, we can directly copy data from user buffer to the socket buffer, instead of creating temporary kernel buffers and do two redundant data copy. But I can't think up of a suitable abstraction for smoltcp to accept user buffer. If you have some idea, we can open a seperate issue to discuss about it.